### ISSUE TITLE
UTF-16 decoder can panic on invalid input

### ISSUE THREAD

**Issue by vickenty** [hsivonen/encoding_rs#44]
This program panics ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=723805f896afcd92e4fa0c1a303c934a)):

```rust
use encoding_rs::CoderResult;

fn main() {
    let d = &mut encoding_rs::UTF_16BE.new_decoder_without_bom_handling();
    let b = &mut [0; 4];

    assert_eq!(
        d.decode_to_utf8(&[217, 99], b, false),
        (CoderResult::InputEmpty, 2, 0, false)
    );
    
    let _ = d.decode_to_utf8(&[217, 99], b, true);
}
```

```
thread 'main' panicked at 'index out of bounds: the len is 4 but the index is 4', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_rs-0.8.20/src/lib.rs:3884:21
```

This is a minimized example of a bug found by a fuzzer in another crate. Let me know if you want full details.

**Comment by hsivonen** [comment]
Thank you for the report. I've noticed this report, but I been super-busy this week. I will get to this.

**Comment by hsivonen** [comment]
Thank you. Fixed. I'll look at another report before pushing a new release.

**Comment by hsivonen** [comment]
The other bug needs more investigation. This fix is now in 0.8.21 on crates.io.
